core/modules/wbase/SendChannel.cc

```diff
@@ -139,6 +139,23 @@ SendChannel::Ptr SendChannel::newStringChannel(std::string& d) {
 }
 
 
+void SendChannel::setTaskCount(int taskCount) {
+    _taskCount = taskCount;
+}
+
+
+bool SendChannel::transmitTaskLast(bool inLast) {
+    /// _caller must have locked _streamMutex before calling this.
+    if (not inLast) {
+        // This wasn't the last message buffer for this task, so it doesn't matter.
+        return false;
+    }
+    ++_lastCount;
+    bool lastTaskDone = _lastCount >= _taskCount;
+    return lastTaskDone;
+}
```

Comment:
The code could be rewritten to be as short as this:
```
bool SendChannel::transmitTaskLast(bool inLast) {
    /// _caller must have locked _streamMutex before calling this.
    if (not inLast) inLast;  // This wasn't the last message buffer for this task, so it doesn't matter.
    return _lastCount >= ++_taskCount;
}
```
I would also suggest passing the lock as a parameter into the method to indicate that the method requires a valid lock. Then you don't need to make the comment - it will become obvious in both definition and declarations of the method:
```
bool SendChannel::transmitTaskLast(lock_guard<mutex> const& lock, bool inLast) {
    if (not inLast) inLast;  // This wasn't the last message buffer for this task, so it doesn't matter.
    return _lastCount >= ++_taskCount;
}

Reply:
I'll shorten it up a bit but I'm leaving lastTaskDone in there as it describes the return value without needing a comment.
Generally, I don't think forcing the mutex to be passed as an argument is that helpful, but the use of the mutex is in this class is weird enough to merit it.

Reply:
I added a SendChannelShared class as adding lock_guard<mutex> const& lock to one function is pointless if it isn't added to all of the functions where it needs to be locked. There are a lot of cases where SendChannel is used and no locking is needed. Polluting those with unneeded locks would be undesirable. 
SendChannelShared holds the mutex as well as other shared only member variables, and has a SendChannel member set at construction.

Reply:
I know this isn't your fault as tis code has been around for "ages" (in Qserv life). But I still find it really scary that the calling the same method from one context may require some external locking while, doing so from some other context - won't. This is one of the things that makes the core hard to debug, and that may result in mysterious crashes. I'm not really suggesting that any significant effort was spent on reinforcing and streamlining locking behavior right now as it may be a significant. However, that needs to be done at some point in a context of some major code overhaul/refactoring project.

Reply:
The SendChannelShared class always requires locking, but the SendChannel class and it's children do not require locking. So I don't really understand the question.